From 0ce55769950ab813a6d1cee378f42b1a606dca1f Mon Sep 17 00:00:00 2001
From: Luke Dashjr <luke-jr+git@utopios.org>
Date: Tue, 26 Jan 2021 22:06:52 +0000
Subject: [PATCH] Update checkpoints and chain params, adding a new checkpoint
 at block 667,811

---
 src/chainparams.cpp | 27 +++++++++++++++++++++------
 1 file changed, 21 insertions(+), 6 deletions(-)

diff --git a/src/chainparams.cpp b/src/chainparams.cpp
index 88cf5ef0a88e7..d6155dd168942 100644
--- a/src/chainparams.cpp
+++ b/src/chainparams.cpp
@@ -91,8 +91,8 @@ class CMainParams : public CChainParams {
         consensus.vDeployments[Consensus::DEPLOYMENT_TAPROOT].nStartTime = 1199145601; // January 1, 2008
         consensus.vDeployments[Consensus::DEPLOYMENT_TAPROOT].nTimeout = 1230767999; // December 31, 2008
 
-        consensus.nMinimumChainWork = uint256S("0x00000000000000000000000000000000000000001533efd8d716a517fe2c5008");
-        consensus.defaultAssumeValid = uint256S("0x0000000000000000000b9d2ec5a352ecba0592946514a92f14319dc2b367fc72"); // 654683
+        consensus.nMinimumChainWork = uint256S("0x0000000000000000000000000000000000000000189ccba50031c45c16a96544");
+        consensus.defaultAssumeValid = uint256S("0x00000000000000000007888a9d01313d69d6335df46ea33e875ee6832670c596"); // 667811
 
         /**
          * The message start string is designed to be unlikely to occur in normal data.
@@ -158,14 +158,29 @@ class CMainParams : public CChainParams {
                 {250000, uint256S("0x000000000000003887df1f29024b06fc2200b55f8af8f35453d7be294df2d214")},
                 {279000, uint256S("0x0000000000000001ae8c72a0b0c301f67e3afca10e819efa9041e458e9bd7e40")},
                 {295000, uint256S("0x00000000000000004d9b4ef50f0f9d686fd69db2e03af35a100370c64632a983")},
+                {393216, uint256S("0x00000000000000000390df7d2bdc06b9fcb260b39e3fb15b4bc9f62572553924")},
+                {421888, uint256S("0x000000000000000004b232ad9492d0729d7f9d6737399ffcdaac1c8160db5ef6")},
+                {438784, uint256S("0x0000000000000000040d6ef667d7a52caf93d8e0d1e40fd7155c787b42667179")},
+                {451840, uint256S("0x0000000000000000029103c8ade7786e7379623465c72d71d84624eb9c159bea")},
+                {469766, uint256S("0x000000000000000000130b2bd812c6a7ae9c02a74fc111806b1dd11e8975da45")},
+                {481824, uint256S("0x0000000000000000001c8018d9cb3b742ef25114f27563e3fc4a1902167f9893")},
+                {514048, uint256S("0x00000000000000000022fe630be397a62c58972bb81f0a2d1ae8c968511a4659")},
+                {553472, uint256S("0x0000000000000000000e06b6698a4f65ab9915f24b23ca2f9d1abf30cc3e9173")},
+                {571392, uint256S("0x00000000000000000019c18b43077775fc299a6646ab0e9dbbd5770bf6ca392d")},
+                {596000, uint256S("0x0000000000000000000706f93dc673ca366c810f317e7cfe8d951c0107b65223")},
+                {601723, uint256S("0x000000000000000000009837f74796532b21d8ccf7def3dcfcb45aa92cd86b9e")},
+                {617056, uint256S("0x0000000000000000000ca51b293fb2be2fbaf1acc76dcbbbff7e4d7796380b9e")},
+                {632549, uint256S("0x00000000000000000001bae1b2b73ec3fde475c1ed7fdd382c2c49860ec19920")},
+                {643700, uint256S("0x00000000000000000002959e9b44507120453344794df09bd1276eb325ed7110")},
+                {667811, uint256S("0x00000000000000000007888a9d01313d69d6335df46ea33e875ee6832670c596")},
             }
         };
 
         chainTxData = ChainTxData{
-            // Data from RPC: getchaintxstats 4096 0000000000000000000b9d2ec5a352ecba0592946514a92f14319dc2b367fc72
-            /* nTime    */ 1603995752,
-            /* nTxCount */ 582083445,
-            /* dTxRate  */ 3.508976121410527,
+            // Data as of block 00000000000000000007888a9d01313d69d6335df46ea33e875ee6832670c596 (height 667811).
+            /* nTime    */ 1611696126,
+            /* nTxCount */ 610018836,
+            /* dTxRate  */ 3.8,
         };
     }
 };
